Microsoft/ChakraCore
# JIT: enable JIT on Linux

Splitted into 4 commits.

JIT: enable JIT on Linux

    Primary changes to JIT on Linux.

    Custom calling convention
    -------------------------

    Calls to native helpers conform to Sys V AMD64 ABI.
        - Change all code to use correct arg registers.

    Calls to JsMethod and native helpers that depend on
    ARGUMENTS/RUNTIME_ARGUMENTS use custom calling convention:
        - Caller home arg registers onto stack.

    Implemented all needed assembly thunks.

    Unwind data
    -----------

    Emit .eh_frame data for dynamic interpreter thunks and all jit functions.
        - For less noise, the new code hides behind existing
            PrologEncoder/PData/XData.

    Misc
    ----

    ARGUMENTS/RUNTIME_ARGUMENTS:
        - Extended to support all runtime usages.
        - Limitation: known arg list must end with "callInfo".

    Int32Math:
        - Use builtins to test signed int overflow. Original code does not work
            due to clang optimization.

    TODO
        - AsmJs: missing ArrayBuffer out of bound access recovery.
        - Simd.

JIT: to compile on Linux

    Noisy changes to compile native code gen code on Linux with clang.
    Most changes are:
        - Fix missing parentheses around `&&` within `||` expressions
        - Fix `char16/_u` issues
        - Fix incorrect `inline`s


JIT: PAL related changes

    PAL related changes to build JIT on Linux.
        - pal.h: bug fixes in BitScan.
        - Add missing APIs, enable Event.


JIT: build and test changes

    Build:
        - Add backend files.
        - Enable --no-jit option, and as default on OSX (temporarily).

    Test:
        - Add `dynapogo` test variant. Change `interpreted` test variant to
            create dynamic profiles and input to `dynapogo` test variant.
            To ensure the same profile file is used for these 2 variants,
            TestLoader (new) assigns each test case a unique test id, and uses
            the id to generate dynamic profile file name.
        - Add 'disable_jit` test variant and as default on OSX (temporarily).
        - Exclude 2 failing tests that need further work.
